src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -61,6 +69,11 @@ internal sealed partial class ObjectReader : IDisposable
             _reader = new BinaryReader(stream, Encoding.UTF8);
             _objectReferenceMap = new ReaderReferenceMap<object>();
             _stringReferenceMap = new ReaderReferenceMap<string>();
+
+            // Capture a copy of the current static binder state.  That way we don't have to 
+            // access any locks while we're doing our processing.
+            _binderSnapshot = ObjectBinder.GetSnapshot();
```

Comment:
StreamReading/Writing is not defined for that sort of operation.  The intent is that you must have registered all your readers ahead of time before serializing out your item.  This is necessarily the case for everything we serialize.  That's because all the objects that we serialize register their types in their static constructor.   i.e. you couldn't get an instance of the type in the first place, without it already having registered the reader in the binder.

Note that this is not a general purpose serialization library.  This is for roslyn to specifically serialize out its own data as quickly as possible.  So this is just a requirement of that system.  I can comment as such.